### PR TITLE
wcs.to_header() precision

### PR DESCRIPTION
**Pull request by nden**
`wcslib-5.14` defines new output [precision constants](https://github.com/astropy/astropy/blob/master/cextern/wcslib/C/wcshdr.h#L975). This PR adds them to `astropy.wcs`. The rules are explained in the link and are not straightforward as the output formatting depends on the input and if the input values differ by orders of magnitude, there may be loss of precision in the output.

Updated to match the implementation:

A default output precision of `22.14G` was set in `to_header`. 
Also the WCSHDO_SIP flag was internally updated as it is in conflict now with another wcslib internal flag `WCSHDO_P16`.

@astrofrog  @eteq @keflavich

### CODE REVIEW

**Comment by astrofrog** [comment]
@nden - thanks for doing this. Is there a way we can default to maximum precision in Astropy?

**Comment by astrofrog** [comment]
In other words, I don't think the tests should be updated to have a different precision from the default, otherwise we're not testing the default behavior, which is where the issue was.

**Comment by nden** [comment]
@astrofrog This is the default behaviour in wcslib. We can set a different default in astropy.wcs by setting one of the flags as default. To some extent the precision depends on the input, though, and the default we set may not always work.  Is `P12` good as a default? It is the default precision for non-parametric keywords. 
Also I noticed there's a conflict with the WCSHDO_SIP bit flag we set in wcs.py. This will have to be changed as well.

**Comment by astrofrog** [comment]
@nden - just to make sure I understand, the default for keywords such as CRVAL1 is WCSHDO_P17 or WCSHDO_P12?

File: docs/wcs/relax.rst [hunk 1]
```diff
@@ -418,3 +418,15 @@ The flag bits are:
   potentially unsafe and is not recommended at this time.
 
 - ``WCSHDO_SIP``: Write out Simple Imaging Polynomial (SIP) keywords.
+
+- ``WCSHDO_P12``, ``WCSHDO_P13``, ``WCSHDO_P14``, ``WCSHDO_P15``, ``WCSHDO_P16``, ``WCSHDO_P17``, ``WCSHDO_EFMT``
+
+  These constants control the precision of the WCS keywords returned by `to_header`.
```
Comment (astrofrog) [inline docs/wcs/relax.rst]: @nden - should be ``~astropy.wcs.WCS.to_header`` otherwise the sphinx build fails because of a warning

**Comment by eteq** [comment]
FWIW, I think P12 is a fine default.  It would be better if there were just an option for "whatever the maximum is", but c'est la vie...

**Comment by astrofrog** [comment]
P12 is not maximum floating-point precision though, so shouldn't we prefer a level of precision that minimizes any precision loss? (P17 for instance)

**Comment by nden** [comment]
I think CRVALs are reported by default in P14 precision. 
I'm going to look now into the problem with one of the flags having the same value as WCSHDO_SIP in wcs.py.

**Comment by astrofrog** [comment]
Ok, P14 is probably a good default, so that works for me.

**Comment by astrofrog** [comment]
@nden - thanks for this! Is it ready for review/merging?

**Comment by nden** [comment]
yes, it's ready for review/merge.

**Comment by eteq** [comment]
Looks good to me.

One possible addition: In #4614 I noted that the current behavior gives different answers for different comibnations of ``CRVAL1`` and ``CRVAL3``.  Maybe just quick add in the example I gave in [this comment](https://github.com/astropy/astropy/pull/4614#issuecomment-186366367) to make sure there's not a regression there?  It really seems like this should fix that, but would be good to make sure...

**Comment by nden** [comment]
@eteq All examples in #4614 work now with this PR. I think the difference is that instead of wcslib trying to figure out what precision to use, it uses the default `P14` now for all WCS keywords.

**Comment by astrofrog** [comment]
I'm going to go ahead and merge this, and @eteq - if you would like to add your examples as further regression tests, you could do that in a separate PR?

**Pull request merged**